packages/filebrowser/src/listing.ts

```diff
@@ -1123,24 +1123,68 @@ export class DirListing extends Widget {
     if (!files || files.length === 0) {
       return;
     }
-    const length = event.dataTransfer?.items.length;
-    if (!length) {
-      return;
-    }
-    for (let i = 0; i < length; i++) {
-      let entry = event.dataTransfer?.items[i].webkitGetAsEntry();
-      if (entry.isDirectory) {
-        console.log('currently not supporting drag + drop for folders');
-        void showDialog({
-          title: this._trans.__('Error Uploading Folder'),
-          body: this._trans.__(
-            'Drag and Drop is currently not supported for folders'
-          ),
-          buttons: [Dialog.cancelButton({ label: this._trans.__('Close') })]
-        });
+    event.preventDefault();
+
+    if (event.dataTransfer && event.dataTransfer.items) {
+      let items = event.dataTransfer.items;
+
+      const addDirectory = (item: any, path: string) => {
+        if (item.isDirectory) {
+          this._model.manager
+            .newUntitled({
+              path: path,
+              type: 'directory'
+            })
+            .then(async model => {
+              await this._manager.rename(
+                `${path}/${model.name}`,
+                `${path}/${item.name}`
+              );
+              this._model
+                .cd(`${path}/${item.name}`)
+                .then(() => {
+                  let directoryReader = item.createReader();
+
+                  directoryReader.readEntries((entries: any) => {
+                    entries.forEach((entry: any) => {
+                      addDirectory(entry, this._model.path);
+                    });
+                  });
```

Comment:
One think to consider here is that the manager is statful, so we need to wait until each directory gets uploaded before doing another thing. I would suggest making the function `addDirectory` `async` and awaiting in this place. This might be easier using a for loop rather than a `forEach`.

Reply:
I changed my `addDir` function as below:
```
if (event.dataTransfer && event.dataTransfer.items) {
      let items = event.dataTransfer.items;

      const addDirectory = async (item: any, path: string) => {
        if (item.isDirectory) {
          this._model.manager
            .newUntitled({
              path: path,
              type: 'directory'
            })
            .then(async model => {
              await this._manager.rename(
                `${path}/${model.name}`,
                `${path}/${item.name}`
              );
              this._model
                .cd(`${path}/${item.name}`)
                .then(() => {
                  let directoryReader = item.createReader();

                  directoryReader.readEntries((entries: any) => {
                    for(let i = 0; i < entries.length; i++) {
                      void addDirectory(entries[i], this._model.path)
                    }
                    // entries.forEach(async (entry: any) => {
                    //   await addDirectory(entry, this._model.path);
                    // });
                  });
                })
                .catch(error =>
                  showErrorMessage(
                    this._trans._p('showErrorMessage', 'Open directory'),
                    error
                  )
                );
            })
            .catch(err => {
              console.log('error while creating folder: ', err);
            });

          // let directoryReader = item.createReader();

          // directoryReader.readEntries((entries: any) => {
          //   entries.forEach((entry: any) => {
          //     addDirectory(entry);
          //   });
          // });
        } else {
          item.file((file: any) => {
            void this._model.upload(file);
          });
        }
      };

      for (let i = 0; i < items.length; i++) {
        let item = items[i].webkitGetAsEntry();

        if (item) {
          const addDir = async() => {
            await addDirectory(item, this._model.path);
          }
          void addDir()
        }
      }
```

And these are the error logs in the terminal:
![image](https://user-images.githubusercontent.com/56908732/126446628-5525870c-44a5-4491-a1a6-75b97054d7c1.png)